dsa-23au/java-dsa/NMac-app/src/BubbleSortCSV.java

```diff
@@ -0,0 +1,96 @@
+// The basic idea behind bubble sort is to repeatedly step through the list,
+// compare adjacent items, and swap them if they are in the wrong order.
+// This process is repeated until no more swaps are needed, which means the list is sorted.
+
+
+package bubbleSortCSV;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BubbleSortCSV {
+
+    // Class to represent each driving record headers from the CSV file
+    public static class DrivingRecord {
+        String time;
+        String date;
+        String elapsed;
+        String orig;
+        String dest;
+        String detourEnroute;
+        String driver;
+        String distance;
+
+        // Overriding toString to print each of the driving record columns
+        @Override
+        public String toString() {
+            return time + ", " + date + ", " + elapsed + ", " + orig + ", " + dest + ", " + detourEnroute + ", " + driver + ", " + distance;
+        }
+    }
+
+
+    // Main method of the program
+    public static void main(String[] args) {
+        // Read CSV file and get a list of driving records
+        List<DrivingRecord> records = readCSV("Time_Driving_Spreadsheet_test2.csv");
+
+        // Sort the records based on time
+        bubbleSort(records);
+
+        // Print sorted records
+        System.out.println("Sorted Records by Time:");
+        for (DrivingRecord record : records) {
+            System.out.println(record);
+        }
+    }
+
+    // Method to read the CSV and return a list of DrivingRecord
+    public static List<DrivingRecord> readCSV(String filename) {
+        List<DrivingRecord> records = new ArrayList<>();
+        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
+            String line = br.readLine(); // Skip the header of the CSV
+            while ((line = br.readLine()) != null) {
+                String[] values = line.split(",");
+                DrivingRecord record = new DrivingRecord();
+                // Assign values from CSV to the DrivingRecord object
+                record.time = values[0];
+                record.date = values[1];
+                record.elapsed = values[2];
+                record.orig = values[3];
+                record.dest = values[4];
+                record.detourEnroute = values[5];
+                record.driver = values[6];
+                record.distance = values[7];
+                // Add record to the list
+                records.add(record);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return records;
+    }
+
+    // Method to sort the list of DrivingRecord using BubbleSort algorithm
+    public static void bubbleSort(List<DrivingRecord> list) {
+        int n = list.size();
+        for (int i = 0; i < n - 1; i++) {
+            for (int j = 0; j < n - i - 1; j++) {
```

Comment:
This looks promising. Have you tested it?
Why does `j` stop at `n - i - 1`, which decreases by 1 each time we increment `i` in the outer loop?

Reply:
I have tested it, however since I have moved the code from eclipse to itelliJ, it now has an issue reading the file. I think I need to re-create the package file.
j stops at n - i - 1 because of how bubble sort algorithm and how it functions.
The n - i - 1 expression is an optimization of the bubble sort. It recognizes that with each complete pass through the list the largest unsorted element "bubbles up" to its correct position at the end of the list.
After the first pass, the largest element is guaranteed to be in the last position (n-1). After the second pass, the second-largest element will be in the second to last position (n-2), and so on.